spring-kafka/src/test/java/org/springframework/kafka/listener/TransactionalContainerTests.java

```diff
@@ -199,6 +203,65 @@ public void testConsumeAndProduceTransactionRollback() throws Exception {
 		verify(pf, times(1)).createProducer();
 	}
 
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	@Test
+	public void testConsumeAndProduceTransactionExternalTM() throws Exception {
+		Consumer consumer = mock(Consumer.class);
+		final TopicPartition topicPartition = new TopicPartition("foo", 0);
+		willAnswer(i -> {
+			((ConsumerRebalanceListener) i.getArgument(1))
+					.onPartitionsAssigned(Collections.singletonList(topicPartition));
+			return null;
+		}).given(consumer).subscribe(any(Collection.class), any(ConsumerRebalanceListener.class));
+		final ConsumerRecords records = new ConsumerRecords(Collections.singletonMap(topicPartition,
+				Collections.singletonList(new ConsumerRecord<>("foo", 0, 0, "key", "value"))));
+		final AtomicBoolean done = new AtomicBoolean();
+		willAnswer(i -> {
+			if (done.compareAndSet(false, true)) {
+				return records;
+			}
+			else {
+				Thread.sleep(500);
+				return null;
+			}
+		}).given(consumer).poll(anyLong());
+		ConsumerFactory cf = mock(ConsumerFactory.class);
+		willReturn(consumer).given(cf).createConsumer("group", null);
+		Producer producer = mock(Producer.class);
+		final CountDownLatch commitLatch = new CountDownLatch(1);
+		willAnswer(i -> {
+			commitLatch.countDown();
+			return null;
+		}).given(producer).commitTransaction();
+		final ProducerFactory pf = mock(ProducerFactory.class);
+		given(pf.transactionCapable()).willReturn(true);
+		given(pf.createProducer()).willReturn(producer);
+		ContainerProperties props = new ContainerProperties("foo");
+		props.setGroupId("group");
+		props.setTransactionManager(new SomeOtherTransactionManager());
+		final KafkaTemplate template = new KafkaTemplate(pf);
+		props.setMessageListener((MessageListener<String, String>) m -> {
+			template.send("bar", "baz");
+			template.sendOffsetsToTransaction(Collections.singletonMap(new TopicPartition(m.topic(), m.partition()),
```

Comment:
The problem is that when the TXM is not Kafka, we don't have a reference to the `ProducerFactory` which is the key for the bound resource. We need that resource to get access to the producer to send the offsets.